client/callgrind_converter.cpp

```diff
@@ -1,37 +1,84 @@
 #include "callgrind_converter.hpp"
 
+#include <algorithm>
+#include <iostream>
+#include <queue>
+
 namespace HawkTracer
 {
 namespace client
 {
 
 CallgrindConverter::~CallgrindConverter()
 {
-    uninit();
 }
 
 bool CallgrindConverter::init(const std::string& file_name)
 {
-    _file.open(file_name);
-    if (_file.is_open())
+    _file_name = file_name;
+    return true;
+}
+
+void CallgrindConverter::process_event(const parser::Event& event)
+{
+    std::string label = _get_label(event);
+
+    if (label == "")
     {
-        _file << "# callgrind format\n";
-        return true;
+        return;
     }
-    return false;
+    HT_ThreadId thread_id = event.get_value_or_default<HT_ThreadId>("thread_id", 0);
+    HT_TimestampNs start_ts = event.get_value<HT_TimestampNs>("timestamp");
+    HT_DurationNs duration = event.get_value_or_default<HT_DurationNs>("duration", 0u);
+    _events[thread_id].emplace_back(label, start_ts, duration);
 }
 
-
-void CallgrindConverter::uninit()
+void CallgrindConverter::_print_function(std::ofstream& file, std::shared_ptr<CallGraph::TreeNode> root)
 {
-    if (_file.is_open())
+    std::queue<std::pair<std::shared_ptr<CallGraph::TreeNode>, std::string>> fnc_queue;
+    fnc_queue.emplace(root, root->data.label + "()");
+
+    while (!fnc_queue.empty())
     {
-        _file.close();
+        auto& fnc = fnc_queue.front();
+        file << "fn=" << fnc.second << "\n";
+        file << "1 " << fnc.first->total_duration - fnc.first->total_children_duration << "\n";
+        for (const auto& child : fnc.first->children)
+        {
+            std::string child_label = child.first->data.label + "()'" + fnc.second;
+            file << "cfn=" << child_label << "\n";
+            file << "calls=" << child.second << " 1\n";
+            file << "1 " << child.first->total_duration << "\n";
+            fnc_queue.emplace(child.first, child_label);
+        }
+        fnc_queue.pop();
     }
 }
 
-void CallgrindConverter::process_event(const parser::Event& event)
+void CallgrindConverter::stop()
 {
+    for (auto& thread : _events)
+    {
+        CallGraph _call_graph;
+        auto root_calls = _call_graph.make(thread.second);
+
+        std::ofstream thread_output_file(_file_name + "." + std::to_string(thread.first));
```

Comment:
You use the file_name twice; here and in the error message. Consider assigning it to variable.